### PR TITLE
[feature](nereids) Add ssb related expressions and PlanNode

### PR DESCRIPTION
**Pull request by zhengshiJ**
# Proposed changes

Add related expressions and AggPlan and SortPlan.

This PR includes the addition of the following expression types:
Arithmetic/BetweenPredicate/CompoundPredicate/FunctionCall.

PlanNode:
LogicalAggregation/LogicalSort

Add a ut to verify related expressions and node parsing

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -178,6 +221,12 @@ booleanValue
     : TRUE | FALSE
     ;
 
+aggFunction
+    : AVG '(' DISTINCT? expression ')'
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:225]: Is this a temporary solution? it is better not add any function name here.
If this is a temporary solution, please add a todo comment.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Logical Aggregation plan operator.
+ */
+public class LogicalAggregation extends LogicalUnaryOperator {
+
+    private final List<? extends NamedExpression> aggregation;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:37]: i think we need to add group by expression here

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    public List<Slot> computeOutput(Plan input) {
+        return sortItems.stream()
+                .map(namedExpr -> {
+                    try {
+                        return namedExpr.getSort().toSlot();
+                    } catch (UnboundException e) {
+                        throw new IllegalStateException(e);
+                    }
+                })
+                .collect(ImmutableList.toImmutableList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java:62]: sort node's output should just return child's output, but not use sort item

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BetweenPredicate.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.doris.nereids.trees.NodeType;
+import org.apache.doris.nereids.types.BooleanType;
+import org.apache.doris.nereids.types.DataType;
+
+/**
+ * Between predicate expression.
+ */
+public class BetweenPredicate extends Expression {
+
+    private boolean isNotBetween;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BetweenPredicate.java:31]: we should not include NOT here, but use NotExpression with BetweenPredicate as child

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/analysis/FunctionParams.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Function's Params.
+ */
+public class FunctionParams {
+    private boolean isStar;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/analysis/FunctionParams.java:31]: just curious, why FunctionParams is a top level class, but SortItem is an inner class?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Arithmetic.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        DIVIDE("/", "divide", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.DIVIDE),
+        MOD("%", "mod", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.MOD),
+        INT_DIVIDE("DIV", "int_divide", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.INT_DIVIDE),
+        ADD("+", "add", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.ADD),
+        SUBTRACT("-", "subtract", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.SUBTRACT),
+        BITAND("&", "bitand", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.BITAND),
+        BITOR("|", "bitor", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.BITOR),
+        BITXOR("^", "bitxor", Arithmetic.OperatorPosition.BINARY_INFIX, TExprOpcode.BITXOR),
+        BITNOT("~", "bitnot", Arithmetic.OperatorPosition.UNARY_PREFIX, TExprOpcode.BITNOT),
+        FACTORIAL("!", "factorial", Arithmetic.OperatorPosition.UNARY_POSTFIX, TExprOpcode.FACTORIAL);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Arithmetic.java:51]: these maybe a UnaryExpression? For the more, based on the previous discussion, we should break them down into separate expressions, should we @EmmyMiao87 ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Literal.java [hunk 7]
```diff
@@ -109,7 +109,7 @@ public <R, C> R accept(ExpressionVisitor<R, C> visitor, C context) {
 
     @Override
     public String sql() {
-        return null;
+        return value.toString();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Literal.java:112]: Should we break them down into separate expressions @EmmyMiao87 ?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 8]
```diff
@@ -132,11 +153,18 @@ private LogicalPlan plan(ParserRuleContext tree) {
     public LogicalPlan visitQuery(QueryContext ctx) {
         Supplier<LogicalPlan> f = () -> {
             // TODO: need to add withQueryResultClauses and withCTE
-            return plan(ctx.queryTerm());
+            LogicalPlan query = plan(ctx.queryTerm());
+            LogicalPlan queryOrganization = withQueryOrganization(ctx.queryOrganization(), query);
+            return queryOrganization == null ? query : queryOrganization;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:158]: should we move this statement into withQueryOrganization

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 9]
```diff
@@ -132,11 +153,18 @@ private LogicalPlan plan(ParserRuleContext tree) {
...
         };
         return ParserUtils.withOrigin(ctx, f);
     }
 
+    private LogicalPlan withQueryOrganization(QueryOrganizationContext ctx, LogicalPlan children) {
+        List<SortItems> sortItems = visitQueryOrganization(ctx);
+        return sortItems == null ? null : new LogicalUnaryPlan(new LogicalSort(sortItems), children);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:165]: maybe it is better
```suggestion
        return sortItems == null ? children : new LogicalUnaryPlan(new LogicalSort(sortItems), children);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 10]
```diff
@@ -162,7 +191,7 @@ public List<Expression> visitNamedExpressionSeq(NamedExpressionSeqContext ctx) {
         List<Expression> expressions = Lists.newArrayList();
         if (ctx != null) {
             for (NamedExpressionContext namedExpressionContext : ctx.namedExpression()) {
-                NamedExpression namedExpression = typedVisit(namedExpressionContext);
+                Expression namedExpression = typedVisit(namedExpressionContext);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:194]: 👍🏻

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 (hunk 1 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:227]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java (hunk 2 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java (hunk 3 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 8 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 9 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 11]
```diff
@@ -278,6 +317,70 @@ last, plan(join.relationPrimary())
...
+    /**
+     * Generate sortItems.
+     *
+     * @param ctx SortItemContext
+     * @return SortItems
+     */
+    public SortItems genSortItems(SortItemContext ctx) {
+        boolean orderDirection = true;
+        if (ctx.DESC() != null) {
+            orderDirection = false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:351]: maybe a enum is better than a boolean

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 12]
```diff
@@ -278,6 +317,70 @@ last, plan(join.relationPrimary())
...
+        List<NamedExpression> groupByExpressions = tmpExpressions.stream().map(expression -> {
+            if (expression instanceof NamedExpression) {
+                return (NamedExpression) expression;
+            } else {
+                return new UnboundAlias(expression);
+            }
+        }).collect(Collectors.toList());
+        List<Expression> aggExpressions = new ArrayList<>();
+        for (Expression expression : expressions) {
+            if (expression instanceof FunctionCall) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:335]: should put all expression to aggExpressions, not only function call

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Override
+    public String toString() {
+        return "Aggregation (" + StringUtils.join(groupByExpressions, ", ")
+                + StringUtils.join(aggExpressions, ", ") + ")";
+    }
+
+    @Override
+    public List<Slot> computeOutput(Plan input) {
+        return groupByExpressions.stream()
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:77]: use agg expressions instead of group by

Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public String toString() {
+        return "Aggregation (" + StringUtils.join(groupByExpressions, ", ")
+                + StringUtils.join(aggExpressions, ", ") + ")";
+    }
+
+    @Override
+    public List<Slot> computeOutput(Plan input) {
+        return aggExpressions.stream()
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:76]: Output should be the agg expr + group by expr

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    private final List<? extends NamedExpression> groupByExpressions;
+    private final List<? extends NamedExpression> aggExpressions;
+
+    /**
+     * Desc: Constructor for LogicalAggregation.
+     */
+    public LogicalAggregation(List<? extends NamedExpression> groupByExpressions,
+            List<? extends NamedExpression> aggExpressions) {
+        super(OperatorType.LOGICAL_AGGREGATION);
+        this.groupByExpressions = Objects.requireNonNull(groupByExpressions, "groupByExpressions can not be null");
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:46]: currently we use empty list to do that

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java [hunk 16]
```diff
@@ -35,6 +35,8 @@ public enum OperatorType {
     LOGICAL_FILTER,
     LOGICAL_JOIN,
     PLACE_HOLDER,
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java:37]: > Put PLACE_HOLDER in the end

place holder has been removed by #10241

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 17]
```diff
@@ -218,7 +251,14 @@ private LogicalPlan visitCommonSelectQueryClausePlan(
             withProject = withFilter;
         }
 
-        return withProject;
+        LogicalPlan withAgg;
+        if (aggClause != null) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:255]: we should do with project only when aggClause is null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CompoundPredicate.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+package org.apache.doris.nereids.trees.expressions;
+
+import org.apache.doris.nereids.rules.expression.rewrite.ExpressionVisitor;
+import org.apache.doris.nereids.trees.NodeType;
+
+/**
+ * Compound predicate expression.
+ * Such as &&,||,AND,OR.
+ */
+public class CompoundPredicate<LEFT_CHILD_TYPE extends Expression, RIGHT_CHILD_TYPE extends Expression>
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CompoundPredicate.java:27]: According to what we discussed before, in terms of expressions, try not to use the combination + inheritance method but directly use the inheritance method.
So I suggest that there should be `and`, `or` two separate expr classes and unified inheritance from `compoundpredicate`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 19]
```diff
@@ -162,7 +192,7 @@ public List<Expression> visitNamedExpressionSeq(NamedExpressionSeqContext ctx) {
         List<Expression> expressions = Lists.newArrayList();
         if (ctx != null) {
             for (NamedExpressionContext namedExpressionContext : ctx.namedExpression()) {
-                NamedExpression namedExpression = typedVisit(namedExpressionContext);
+                Expression namedExpression = typedVisit(namedExpressionContext);
                 expressions.add(namedExpression);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:196]:
```suggestion
                Expression expression = typedVisit(namedExpressionContext);
                expressions.add(expression);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Logical Aggregation plan operator.
+ */
+public class LogicalAggregation extends LogicalUnaryOperator {
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java:42]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.doris.nereids.trees.plans.Plan;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+    Logical Sort plan operator.
+ */
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalSort.java:38]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 17 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:259]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CompoundPredicate.java (hunk 18 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/CompoundPredicate.java:26]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Literal.java (hunk 7 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Literal.java:112]: Added todo information, which is subsequently rewritten in subexpressions

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+
+    private final List<? extends NamedExpression> groupByExpressions;
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/logical/LogicalAggregation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 23]
```diff
@@ -278,6 +318,64 @@ last, plan(join.relationPrimary())
...
+    public SortItems genSortItems(SortItemContext ctx) {
+        OrderDirection orderDirection;
+        if (ctx.DESC() != null) {
+            orderDirection = OrderDirection.DESC;
+        } else {
+            orderDirection = OrderDirection.ASC;
+        }
+        Expression expression = typedVisit(ctx.expression());
+        NamedExpression namedExpression;
+        if (expression instanceof NamedExpression) {
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 19 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

**Review by morrySnow — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**